File: app/src/main/java/ru/profapp/ranobereader/ChapterRecyclerViewAdapter.java
package ru.profapp.ranobereader;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;

import ru.profapp.ranobereader.DAO.Database;
import ru.profapp.ranobereader.Models.Chapter;
import ru.profapp.ranobereader.Models.TextChapter;

public class ChapterRecyclerViewAdapter extends
        RecyclerView.Adapter<ChapterRecyclerViewAdapter.ViewHolder> {

    private final List<Chapter> mValues;
    Drawable downloadImage, downloadDoneImage;

    public ChapterRecyclerViewAdapter(List<Chapter> items) {
        mValues = items;
    }

    public Drawable getDownloadImage() {
        return downloadImage;
    }

    public void setDownloadImage(Drawable downloadImage) {
        this.downloadImage = downloadImage;
    }

    public Drawable getDownloadDoneImage() {
        return downloadDoneImage;
    }

    public void setDownloadDoneImage(Drawable downloadDoneImage) {
        this.downloadDoneImage = downloadDoneImage;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.chapter_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.mItem = mValues.get(position);
        holder.mIdView.setText(mValues.get(position).getTitle());

        holder.mView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                Intent intent = new Intent(v.getContext(), ChapterText.class);
                intent.putExtra("ChapterInfo", holder.mItem);
                v.getContext().startActivity(intent);
            }
        });

        holder.mImageButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final TextChapter textChapter = new TextChapter(holder.mItem.getUrl(),
                        getText(holder.mItem, holder.mContext));
                AsyncTask.execute(new Runnable() {
                    @Override
                    public void run() {
                        Database.getInstance(holder.mView.getContext()).getTextDao().insert(textChapter);
                    }
                });
                holder.mImageButton.setImageDrawable(downloadDoneImage);

            }
        });
        if (! holder.mItem.getCanRead()){
            holder.mIdView.setBackgroundColor(Color.GRAY);
        }

    }

    private String getText(Chapter chapter, Context context) {
        ChapterText chapterText = new ChapterText();
        chapterText.GetChapterText(chapter, context);
        return chapterText.mChapter.getText();

    }

    @Override
    public int getItemCount() {
        return mValues.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        final View mView;
        final TextView mIdView;
        final ImageButton mImageButton;
        Chapter mItem;
        final Context mContext;


        ViewHolder(View view) {
            super(view);
            mView = view;
            mIdView = (TextView) view.findViewById(R.id.id);
            mContext = view.getContext();
            mImageButton = (ImageButton) view.findViewById(R.id.imageButton_download);
        }

    }
}

File: app/src/main/java/ru/profapp/RanobeReader/Adapters/ChapterRecyclerViewAdapter.kt
package ru.profapp.RanobeReader.Adapters

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.NonNull
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import ru.profapp.RanobeReader.Activities.ChapterTextActivity
import ru.profapp.RanobeReader.Common.Constants
import ru.profapp.RanobeReader.Models.Chapter
import ru.profapp.RanobeReader.Models.Ranobe
import ru.profapp.RanobeReader.MyApp
import ru.profapp.RanobeReader.R

class ChapterRecyclerViewAdapter(private val context: Context, private val mValues: List<Chapter>, private val mRanobe: Ranobe) : RecyclerView.Adapter<ChapterRecyclerViewAdapter.ViewHolder>() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)

    private val clickListener = object : OnItemClickListener {
        override fun onItemClick(item: Chapter) {
            if (item.canRead) {
                if (MyApp.ranobe == null || item.ranobeUrl != MyApp.ranobe!!.url) {

                    var ranobe = Ranobe()
                    ranobe.url = item.ranobeUrl
                    if (MyApp.fragmentType != null && MyApp.fragmentType != Constants.FragmentType.Saved) {
                        try {
                            ranobe = ranobe.updateRanobe(context).blockingGet()
                        } catch (ignored: Exception) {
                            ranobe = mRanobe
                        }

                    } else {
                        ranobe = mRanobe
                    }

                    MyApp.ranobe = ranobe
                }
                if (MyApp.ranobe != null) {
                    val intent = Intent(context, ChapterTextActivity::class.java)
                    intent.putExtra("ChapterIndex", item.index)
                    context.startActivity(intent)
                }
            }

        }

    }

    interface OnItemClickListener {
        fun onItemClick(item: Chapter)
    }


    @NonNull
    override fun onCreateViewHolder(@NonNull parent: ViewGroup, viewType: Int): ChapterRecyclerViewAdapter.ViewHolder {

        val view = inflater.inflate(R.layout.item_chapter, parent, false)
        return ViewHolder(view)

    }


    override fun onBindViewHolder(@NonNull holder: ChapterRecyclerViewAdapter.ViewHolder, position: Int) {

        holder.mChapterItem = mValues[position]
        holder.mTextView.text = mValues[position].title

        if (!holder.mChapterItem.canRead) {
            holder.mView.setBackgroundColor(Color.GRAY)
        }

        if (holder.mChapterItem.isRead) {
            holder.mView.setBackgroundColor(ContextCompat.getColor(context, R.color.colorPrimaryDark))
        }


    }

    override fun getItemCount(): Int {
        return mValues.size
    }

    inner class ViewHolder(val mView: View) : RecyclerView.ViewHolder(mView) {
        val mTextView: TextView = mView.findViewById(R.id.id)
        lateinit var mChapterItem: Chapter

        init {

            itemView.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    clickListener.onItemClick(mChapterItem)
                }
            }

        }

    }

}
